scanner/decisions/0012-reindexing.md

```diff
@@ -0,0 +1,186 @@
+into our Indexer logic.
+
+For example, it may be done via the following:
+
+```go
+package example
+
+import (
+    "context"
+    "log"
+
+    "github.com/quay/claircore"
+    "github.com/quay/claircore/indexer"
+    "github.com/quay/claircore/libindex"
+)
+
+func Example() {
+    // Assuming there is already some instance of libindex.Libindex and a claircore.Digest
+    var i *libindex.Libindex
+    var digest claircore.Digest
+
+    // Get all current versioned scanners. 
+    pscnrs, dscnrs, rscnrs, fscnrs, err := indexer.EcosystemsToScanners(context.Background(), i.Ecosystems)
+    if err != nil {
+        log.Fatal(err)
+    }
+    vscnrs := indexer.MergeVS(pscnrs, dscnrs, rscnrs, fscnrs)
+
+    ok, err := i.Store.ManifestScanned(context.Background(), digest, vscnrs)
+    if err != nil {
+        log.Fatal(err)
+    }
+    if ok {
+        // Image already indexed with latest versioned scanners.
+    } else {
+        // Image was either not previously indexed, or it was with older versioned scanners.
+    }
+}
+```
+
+This method does not require any kind of migration to account for previously indexed images, as the data/tables already have all required the information.
+
+#### Alternative - Tracking Indexer State
+
+Scanner V4 Indexer would maintain a table which tracks each successful Index Report (identified by the report's hash ID) 
+alongside the Scanner V4 Indexer's state (equal to the `etag` sent by Clair).
+
+This way, when clients reach out to Scanner V4 Indexer, it may first check not only if the image was already indexed,
+but also the state of the Indexer at the time of indexing. If they do not match, then we will create the full `*claircore.Manifest`
+and re-index the image. As mentioned in the [Context](#context) section, ClairCore already has some builtin performance optimizations for this.
+Another benefit to this that `*claircore.Manifest` continues to only be generated, as needed.
+
+However, there are cons which outweigh the pros:
+
+* More data for Scanner V4 to track outside ClairCore
+  * This is not a blocker, but it is worth mentioning
+* Handling preexisting Index Reports
+  * This is the challenging aspect, which would block this
+  * There are a few options:
+    * Write a migration
+      * This involves doing essentially the same thing `ManifestScanned` does to determine the "versioned scanners"
+      * This may not be able to be pure PostgeSQL, as the `etag` is generated in Go, so that is an extra challenge
+    * Populate the table upon Index Report fetching
+      * If the entry does not exist in the table (the case for all Index Reports created prior to this release), then create the Index Report and populate the table
+      * This essentially forces a re-indexing of all previously indexed images, whether it is required or not, which may be costly
+
+### Random Deletions
```

Comment:
I have some questions that I’d like to be clarified to better understand the ADR:
1. As of now, is it correct to say there is no way to delete an index report in Scanner v4? For example, even the client cannot request the deletion of an index report.
2. Once an index report is deleted in Scanner v4, is it accurate to say that it will not be regenerated until the client (such as the sensor) specifically requests it?
3. If an index report as old as 1 month is not selected in the random deletion process, does that mean it will be living for more than 1 month and it will not be deleted any more?

Reply:
1. Correct. Scanner V4 does not have this API available. Clair does, though. It is used, for example, by Quay when an image is deleted from the registry. When an image is deleted from Quay, Quay reaches out to Clair to delete the manifest
2. Based on this ADR, yes, that's correct. It'll only be recreated on-demand. That means if we delete an Index Report for an image no longer tracked by StackRox, we will no longer have its data stored in Scanner V4
3. Perhaps it's not completely clear, but the idea is to periodically check the `manifest_metadata` table for expired timestamps. From there, we will delete the manifest/Index Report in Claircore for each expired manifest. If this process fails, the hope is the next check for expired manifests will catch the manifest and try again

Reply:
I was thinking why not we delete all index reports at age of 2 - 4 weeks but since we will periodically check the `manifest_metadata` table for expired timestamps, the random deletion solution should be working fine.

Reply:
yeah the hope is even though each image will be scanned at different times, the uniformly random timestamps will eventually help even out the number of manifests/images which expire for each garbage collection cycle. That way we don't do a mass deletion and mass recreation and overload the CPU/memory/registry all at once

 Unspecified here, but in the associated PR, I set it to check for expired manifests every 6 hours, and for each expired manifest, start a transaction to delete the expired entries, and then delete them in claircore as well. if the claircore deletion fails, then rollback the transaction and try again in another 6 hours